### PR TITLE
fix pbm

### PR DESCRIPTION
**Pull request by EnLAI111**

### CODE REVIEW

File: solvers/python-Celer.py [hunk 1]
```diff
@@ -33,21 +33,22 @@ def set_objective(self, A, reg, y):
...
+        A_op = self.A @ np.ones((len_y, len_y)) @ (self.A.T) / (S @ S)
+        y_new = self.y - A_op @ self.y
+        AL_new = self.A @ L - A_op @ self.A @ L
         self.lasso.max_iter = n_iter
         self.lasso.fit(AL_new, y_new)
         z = self.lasso.coef_.flatten()
-        c = np.mean(self.y - AL @ z)
-        self.u = L @ z + np.linalg.pinv(self.A) @ (np.ones(len_y) * c)
+        c = S @ (self.y - self.A @ L @ z) / (S @ S)
+        self.u = L @ z + c
```
Comment (tomMoral) [inline solvers/python-Celer.py:51]: I think it would be more efficient to use `np.cumsum(z, axis=1)` instead of `L @ z`.

File: solvers/python-DPGD.py [hunk 2]
```diff
@@ -29,7 +27,7 @@ def run(self, callback):
         DA_inv = D @ np.linalg.pinv(self.A)
         v = np.zeros(len_y - 1)
         u = np.zeros(len_y)
-        stepsize = 1 / (np.linalg.norm(DA_inv, ord=2)**2)  # 1/ rho
+        stepsize = 1.99 / (np.linalg.norm(DA_inv, ord=2)**2)  # 1.99 / rho
```
Comment (tomMoral) [inline solvers/python-DPGD.py:30]: Does it go faster?

Comment (EnLAI111) [reply solvers/python-DPGD.py:30]: image
yes, it goes a little faster.

File: solvers/python-ISTA.py [hunk 3]
```diff
@@ -9,8 +10,9 @@ class Solver(BaseSolver):
     """Proximal gradient descent for synthesis formulation."""
     name = 'ISTA'
 
-    stopping_strategy = "callback"
-    # any parameter defined here is accessible as a class attribute
+    stopping_criterion = SufficientProgressCriterion(
+        patience=20, strategy='callback'
```
[the remark this answers is not in the document]
Comment (EnLAI111) [reply solvers/python-ISTA.py:14]: Now I set all the patience=10. That's just me trying to see if it's possible to make these algorithms reach the point of convergence. But in fact, it doesn't change a lot, some of algorithms converge too slowly.

File: objective.py [hunk 4]
```diff
@@ -1,26 +1,73 @@
-import numpy as np
 from benchopt import BaseObjective
+import numpy as np
+from scipy import optimize
```
Comment (tomMoral) [inline objective.py:3]:
```suggestion
from benchopt import BaseObjective, safe_import_context
with safe_import_context() as import_ctx:
    import numpy as np
    from scipy import optimize
```

Comment (tomMoral) [reply objective.py:3]: You need to protect these import to avoid bad interaction with  auto completion.

File: objective.py [hunk 5]
```diff
@@ -1,26 +1,73 @@
...
+                    delta=self.delta, data_fit=self.data_fit)
+
+    def huber(self, R, delta):
+        norm_1 = np.abs(R)
+        loss = np.where(norm_1 < delta,
+                        0.5 * norm_1**2,
+                        delta * norm_1 - 0.5 * delta**2)
+        return np.sum(loss)
+
+    def set_c(self, S, delta):
```
Comment (tomMoral) [inline objective.py:40]: This function does not set anything so it should be called `get_c` or `compute_c`.

File: objective.py [hunk 6]
```diff
@@ -1,26 +1,73 @@
...
+                        S * np.sign(R) * delta)).sum())
+        yS = self.y / S
+        return optimize.golden(f, brack=(min(yS), max(yS)))
+
+    def set_reg_max(self, c):
+        L = np.tri(self.y.shape[0])
+        AL = self.A @ L
+        z = np.zeros(self.y.shape[0])
+        z[0] = c
+        return np.max(abs(self.grad(AL, z)))
```
Comment (tomMoral) [inline objective.py:63]: Same: `get_reg_max` or something like that.

**Review by tomMoral — APPROVED** [review]
A few comments.

Overall, this looks good!

**Comment by tomMoral** [comment]
What is the status of this PR @EnLAI111 ? Should we merge it so it is easier to review the rest of the cahnges/

**Comment by EnLAI111** [comment]
Chambolle-Poke can't work for Huber. There is always a theoretical problem of calculation that I can't solve.

**Comment by agramfort** [comment]
@svaiter would have some time to help on the TV benchmarks?

**Comment by svaiter** [comment]
> > What is the status of this PR @EnLAI111 ? Should we merge it so it is easier to review the rest of the cahnges/
> 
> Chambolle-Poke can't work for Huber. There is always a theoretical problem of calculation that I can't solve.

I am not sure to understand your problem. If you look at (Chambolle and Pock, 2011) in Section 6.2.3, I think there is all ingredient to implement the resolvent for Huberized-TV (1D or 2D), if I am not mistaken.

**Comment by EnLAI111** [comment]
> > > What is the status of this PR @EnLAI111 ? Should we merge it so it is easier to review the rest of the cahnges/
> > 
> > 
> > Chambolle-Poke can't work for Huber. There is always a theoretical problem of calculation that I can't solve.
> 
> I am not sure to understand your problem. If you look at (Chambolle and Pock, 2011) in Section 6.2.3, I think there is all ingredient to implement the resolvent for Huberized-TV (1D or 2D), if I am not mistaken.
> 
> EDIT: BTW, I think it is a bit dangerous to mix TV and TV-Huber for Chambolle-Pock since for TV you expect a 1/k^2 rate whereas the other can be linear depending on the operator A. I think it should be in two different benchmarks

Thanks for your instruction! Yes, I have read this paper, but in our case, it's not ```y - x``` but ```y-Ax```.  My calculation is following, and I'm confused by the circled part.

ChamPoke

I'm sorry, when you mention "TV", do you mean "quadratic loss + TV penalty"? If so, I don't think we should be worried. Because  we have a parameter ```data_fit = ['quadratic', 'huber']``` of objective function. For each ```data_fit```, we compare the convergence speed of different algorithm. So "quadratic loss + TV penalty" won't be compare with "huber loss + TV penalty".

File: solvers/python-Celer-Huber.py [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+    import numpy as np
+    from skglm import GeneralizedLinearEstimator
+    from skglm.datafits import Huber
+    from skglm.penalties import WeightedL1
+    from sklearn.exceptions import ConvergenceWarning
+
+
+class Solver(BaseSolver):
+    """Coordinate descent for synthesis formulation as lasso problem."""
+    name = 'Celer-Huber'
```
Comment (tomMoral) [inline solvers/python-Celer-Huber.py:18]: rename `skglm synthesis`

File: solvers/python-Celer-Huber.py [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+    references = [
+        'M. Massias, A. Gramfort and J. Salmon, ICML, '
+        '"Celer: a Fast Solver for the Lasso with Dual Extrapolation", '
+        'vol. 80, pp. 3321-3330 (2018)'
+    ]
+
+    def skip(self, A, reg, y, c, delta, data_fit):
+        if data_fit == 'quad':
+            return True, "solver does not work with quadratic loss"
+        return False, None
```
Comment (tomMoral) [inline solvers/python-Celer-Huber.py:35]: It would be nice to make it work with the quadratic, just use `Lasso` in this case.

File: solvers/python-Celer-Huber.py [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+        self.A, self.y = A, y
+        self.c = c
+        self.delta = delta
+        self.data_fit = data_fit
+
+        warnings.filterwarnings('ignore', category=ConvergenceWarning)
+        weights = np.ones(self.y.shape[0])
+        weights[0] = 0
+        self.clf = GeneralizedLinearEstimator(
+            Huber(self.delta),
```
Comment (tomMoral) [inline solvers/python-Celer-Huber.py:48]:
```suggestion
            Huber(self.delta) if self.data_fir == "huber" else Quadratic(),
```

with `Qudratic` from `sklglm.data_fits`

File: solvers/python-ISTA.py [hunk 10]
```diff
@@ -9,29 +9,46 @@ class Solver(BaseSolver):
...
-        z = np.zeros(len_y)  # initialisation
-        while callback(L @ z):
-            z = self.st(z - stepsize * AL.T @ (AL @ z - self.y),
+        stepsize = self.alpha / (np.linalg.norm(AL, ord=2)**2)  # alpha / rho
+        # initialisation
+        z = np.zeros(len_y)
+        z[0] = self.c
+        while callback(np.cumsum(z)):
+            z = self.st(z - stepsize * self.grad(AL, z),
                         self.reg * stepsize)
```
Comment (tomMoral) [inline solvers/python-ISTA.py:34]: Add fISTa

**Review by tomMoral — COMMENTED** [review]
some comments from our meeting.
